src/lib/operators/projection.cpp

```diff
@@ -51,17 +51,45 @@ std::shared_ptr<const Table> Projection::_on_execute() {
 
   const auto& input_table = *left_input_table();
 
-  /**
-   * For PQPColumnExpressions, it is possible to forward the input column if the input TableType (References or Data)
-   * matches the output column type (ReferenceSegment or not).
-   */
-  const auto only_projects_columns = std::all_of(expressions.begin(), expressions.end(), [&](const auto& expression) {
+  // Determine the type of the output table: If no input columns are forwarded, i.e., all output columns are newly
+  // generated, the output type is always TableType::Data and all segments are ValueSegments. Otherwise, the output type
+  // depends on the input table's type. If the forwarded columns come from a data table, so are the newly generated
+  // columns. However, we cannot return a table that mixes data and reference segments, so if the forwarded columns are
+  // reference columns, the newly generated columns contain reference segments, too. These point to an internal dummy
+  // table, the projection_result_table. The life time of this table is managed by the shared_ptr in
+  // ReferenceSegment::_referenced_table. The ReferenceSegments created for this use an EntireChunkPosList, so
```

Comment:
Not sure whether that applies here but if I remember correctly, the plan cache points to the PQP which holds a shared_ptr to `_output`. Would that mean that a ProjectionResultTable is currently never destroyed as long as its corresponding plan is in the cache?

If this assumption is correct, this behavior is not an issue of this implementation but it might lead to increased memory footprint which might be interesting for the projects that look at memory consumption, right?

Reply:
> Not sure whether that applies here but if I remember correctly, the plan cache points to the PQP which holds a shared_ptr to _output.

Sometimes, yes. That's broken anyway, see #2178. The output of any intermediary operator should not outlive the query. However, this PR does nothing to change this. Previously, the ValueSegment lived for as long as the output table lived. This is the same now, just with the difference that it is deleted via the ReferenceSegment.